app/services/user-agent-service/datomstore.js

```diff
@@ -147,58 +201,79 @@ export class ProfileDatomStorage {
    * @returns {Promise} a promise that resolves to the place ID.
    */
   visit(url, session, title, now = microtime.now()) {
-    const result = datascript.core.transact_BANG_(this.conn, vector(
+    let assertions = vector(
       vector(DB_ADD, -1, 'page/url', url),
-      vector(DB_ADD, -1, 'page/title', title),    // TODO: handle no title.
       vector(DB_ADD, -1, 'event/visit', -2),
       vector(DB_ADD, -2, 'visit/instant', now)
-    ));
+    );
+
+    if (title) {
+      assertions = mori.conj(assertions,
+        vector(DB_ADD, -1, 'page/title', title)
+      );
+    }
+
+    const result = this.transact(assertions);
 
     // TODO: session.
 
-    return datascript.core.resolve_tempid(result.tempids, -1);
+    return getIn(result, [TEMPIDS, -1]);
   }
 
   getVisits(since = null) {
-    const db = datascript.core.db(this.conn);
-    return datascript.core.index_range(db, 'visit/instant', since, null);
+    return mori.reverse(
+      datascript.core.index_range(
+        this.getDB(), 'visit/instant', since, null));
   }
 
-  // Ordered by last visit, descending.
-  async visited(since = null, limit = 10) {
-    const visitDatoms = this.getVisits(since);
+  getPageDetails(visits) {
+    const query = this.pageDetailsQuery;
+
+    function toJSON(pair) {
+      const [ts, details, page] = pair;
+      return {
+        lastVisited: ts,
+        place: page,
+        uri: mori.get(details, 'page/url'),
+        title: mori.get(details, 'page/title'),
+      };
+    }
+
+    const results = datascript.core.q(query, this.getDB(), visits);
```

Comment:
I would like to have `d = datascript.core; djs = datascript.javascript` throughout.